// @openzeppelin/contracts/build/contracts/AccessManager.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import type {
  BaseContract,
  BigNumberish,
  BytesLike,
  FunctionFragment,
  Result,
  Interface,
  EventFragment,
  AddressLike,
  ContractRunner,
  ContractMethod,
  Listener,
} from "ethers";
import type {
  TypedContractEvent,
  TypedDeferredTopicFilter,
  TypedEventLog,
  TypedLogDescription,
  TypedListener,
  TypedContractMethod,
} from "../../../../../common";

export interface AccessManagerInterface extends Interface {
  getFunction(
    nameOrSignature:
      | "ADMIN_ROLE"
      | "ADMIN_ROLE()"
      | "PUBLIC_ROLE"
      | "PUBLIC_ROLE()"
      | "canCall"
      | "canCall(address,address,bytes4)"
      | "cancel"
      | "cancel(address,address,bytes)"
      | "consumeScheduledOp"
      | "consumeScheduledOp(address,bytes)"
      | "execute"
      | "execute(address,bytes)"
      | "expiration"
      | "expiration()"
      | "getAccess"
      | "getAccess(uint64,address)"
      | "getNonce"
      | "getNonce(bytes32)"
      | "getRoleAdmin"
      | "getRoleAdmin(uint64)"
      | "getRoleGrantDelay"
      | "getRoleGrantDelay(uint64)"
      | "getRoleGuardian"
      | "getRoleGuardian(uint64)"
      | "getSchedule"
      | "getSchedule(bytes32)"
      | "getTargetAdminDelay"
      | "getTargetAdminDelay(address)"
      | "getTargetFunctionRole"
      | "getTargetFunctionRole(address,bytes4)"
      | "grantRole"
      | "grantRole(uint64,address,uint32)"
      | "hasRole"
      | "hasRole(uint64,address)"
      | "hashOperation"
      | "hashOperation(address,address,bytes)"
      | "isTargetClosed"
      | "isTargetClosed(address)"
      | "labelRole"
      | "labelRole(uint64,string)"
      | "minSetback"
      | "minSetback()"
      | "multicall"
      | "multicall(bytes[])"
      | "renounceRole"
      | "renounceRole(uint64,address)"
      | "revokeRole"
      | "revokeRole(uint64,address)"
      | "schedule"
      | "schedule(address,bytes,uint48)"
      | "setGrantDelay"
      | "setGrantDelay(uint64,uint32)"
      | "setRoleAdmin"
      | "setRoleAdmin(uint64,uint64)"
      | "setRoleGuardian"
      | "setRoleGuardian(uint64,uint64)"
      | "setTargetAdminDelay"
      | "setTargetAdminDelay(address,uint32)"
      | "setTargetClosed"
      | "setTargetClosed(address,bool)"
      | "setTargetFunctionRole"
      | "setTargetFunctionRole(address,bytes4[],uint64)"
      | "updateAuthority"
      | "updateAuthority(address,address)"
  ): FunctionFragment;

  getEvent(
    nameOrSignatureOrTopic:
      | "OperationCanceled"
      | "OperationCanceled(bytes32,uint32)"
      | "OperationExecuted"
      | "OperationExecuted(bytes32,uint32)"
      | "OperationScheduled"
      | "OperationScheduled(bytes32,uint32,uint48,address,address,bytes)"
      | "RoleAdminChanged"
      | "RoleAdminChanged(uint64,uint64)"
      | "RoleGrantDelayChanged"
      | "RoleGrantDelayChanged(uint64,uint32,uint48)"
      | "RoleGranted"
      | "RoleGranted(uint64,address,uint32,uint48,bool)"
      | "RoleGuardianChanged"
      | "RoleGuardianChanged(uint64,uint64)"
      | "RoleLabel"
      | "RoleLabel(uint64,string)"
      | "RoleRevoked"
      | "RoleRevoked(uint64,address)"
      | "TargetAdminDelayUpdated"
      | "TargetAdminDelayUpdated(address,uint32,uint48)"
      | "TargetClosed"
      | "TargetClosed(address,bool)"
      | "TargetFunctionRoleUpdated"
      | "TargetFunctionRoleUpdated(address,bytes4,uint64)"
  ): EventFragment;

  encodeFunctionData(
    functionFragment: "ADMIN_ROLE",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "ADMIN_ROLE()",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "PUBLIC_ROLE",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "PUBLIC_ROLE()",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "canCall",
    values: [AddressLike, AddressLike, BytesLike]
  ): string;
  encodeFunctionData(
    functionFragment: "canCall(address,address,bytes4)",
    values: [AddressLike, AddressLike, BytesLike]
  ): string;
  encodeFunctionData(
    functionFragment: "cancel",
    values: [AddressLike, AddressLike, BytesLike]
  ): string;
  encodeFunctionData(
    functionFragment: "cancel(address,address,bytes)",
    values: [AddressLike, AddressLike, BytesLike]
  ): string;
  encodeFunctionData(
    functionFragment: "consumeScheduledOp",
    values: [AddressLike, BytesLike]
  ): string;
  encodeFunctionData(
    functionFragment: "consumeScheduledOp(address,bytes)",
    values: [AddressLike, BytesLike]
  ): string;
  encodeFunctionData(
    functionFragment: "execute",
    values: [AddressLike, BytesLike]
  ): string;
  encodeFunctionData(
    functionFragment: "execute(address,bytes)",
    values: [AddressLike, BytesLike]
  ): string;
  encodeFunctionData(
    functionFragment: "expiration",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "expiration()",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "getAccess",
    values: [BigNumberish, AddressLike]
  ): string;
  encodeFunctionData(
    functionFragment: "getAccess(uint64,address)",
    values: [BigNumberish, AddressLike]
  ): string;
  encodeFunctionData(functionFragment: "getNonce", values: [BytesLike]): string;
  encodeFunctionData(
    functionFragment: "getNonce(bytes32)",
    values: [BytesLike]
  ): string;
  encodeFunctionData(
    functionFragment: "getRoleAdmin",
    values: [BigNumberish]
  ): string;
  encodeFunctionData(
    functionFragment: "getRoleAdmin(uint64)",
    values: [BigNumberish]
  ): string;
  encodeFunctionData(
    functionFragment: "getRoleGrantDelay",
    values: [BigNumberish]
  ): string;
  encodeFunctionData(
    functionFragment: "getRoleGrantDelay(uint64)",
    values: [BigNumberish]
  ): string;
  encodeFunctionData(
    functionFragment: "getRoleGuardian",
    values: [BigNumberish]
  ): string;
  encodeFunctionData(
    functionFragment: "getRoleGuardian(uint64)",
    values: [BigNumberish]
  ): string;
  encodeFunctionData(
    functionFragment: "getSchedule",
    values: [BytesLike]
  ): string;
  encodeFunctionData(
    functionFragment: "getSchedule(bytes32)",
    values: [BytesLike]
  ): string;
  encodeFunctionData(
    functionFragment: "getTargetAdminDelay",
    values: [AddressLike]
  ): string;
  encodeFunctionData(
    functionFragment: "getTargetAdminDelay(address)",
    values: [AddressLike]
  ): string;
  encodeFunctionData(
    functionFragment: "getTargetFunctionRole",
    values: [AddressLike, BytesLike]
  ): string;
  encodeFunctionData(
    functionFragment: "getTargetFunctionRole(address,bytes4)",
    values: [AddressLike, BytesLike]
  ): string;
  encodeFunctionData(
    functionFragment: "grantRole",
    values: [BigNumberish, AddressLike, BigNumberish]
  ): string;
  encodeFunctionData(
    functionFragment: "grantRole(uint64,address,uint32)",
    values: [BigNumberish, AddressLike, BigNumberish]
  ): string;
  encodeFunctionData(
    functionFragment: "hasRole",
    values: [BigNumberish, AddressLike]
  ): string;
  encodeFunctionData(
    functionFragment: "hasRole(uint64,address)",
    values: [BigNumberish, AddressLike]
  ): string;
  encodeFunctionData(
    functionFragment: "hashOperation",
    values: [AddressLike, AddressLike, BytesLike]
  ): string;
  encodeFunctionData(
    functionFragment: "hashOperation(address,address,bytes)",
    values: [AddressLike, AddressLike, BytesLike]
  ): string;
  encodeFunctionData(
    functionFragment: "isTargetClosed",
    values: [AddressLike]
  ): string;
  encodeFunctionData(
    functionFragment: "isTargetClosed(address)",
    values: [AddressLike]
  ): string;
  encodeFunctionData(
    functionFragment: "labelRole",
    values: [BigNumberish, string]
  ): string;
  encodeFunctionData(
    functionFragment: "labelRole(uint64,string)",
    values: [BigNumberish, string]
  ): string;
  encodeFunctionData(
    functionFragment: "minSetback",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "minSetback()",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "multicall",
    values: [BytesLike[]]
  ): string;
  encodeFunctionData(
    functionFragment: "multicall(bytes[])",
    values: [BytesLike[]]
  ): string;
  encodeFunctionData(
    functionFragment: "renounceRole",
    values: [BigNumberish, AddressLike]
  ): string;
  encodeFunctionData(
    functionFragment: "renounceRole(uint64,address)",
    values: [BigNumberish, AddressLike]
  ): string;
  encodeFunctionData(
    functionFragment: "revokeRole",
    values: [BigNumberish, AddressLike]
  ): string;
  encodeFunctionData(
    functionFragment: "revokeRole(uint64,address)",
    values: [BigNumberish, AddressLike]
  ): string;
  encodeFunctionData(
    functionFragment: "schedule",
    values: [AddressLike, BytesLike, BigNumberish]
  ): string;
  encodeFunctionData(
    functionFragment: "schedule(address,bytes,uint48)",
    values: [AddressLike, BytesLike, BigNumberish]
  ): string;
  encodeFunctionData(
    functionFragment: "setGrantDelay",
    values: [BigNumberish, BigNumberish]
  ): string;
  encodeFunctionData(
    functionFragment: "setGrantDelay(uint64,uint32)",
    values: [BigNumberish, BigNumberish]
  ): string;
  encodeFunctionData(
    functionFragment: "setRoleAdmin",
    values: [BigNumberish, BigNumberish]
  ): string;
  encodeFunctionData(
    functionFragment: "setRoleAdmin(uint64,uint64)",
    values: [BigNumberish, BigNumberish]
  ): string;
  encodeFunctionData(
    functionFragment: "setRoleGuardian",
    values: [BigNumberish, BigNumberish]
  ): string;
  encodeFunctionData(
    functionFragment: "setRoleGuardian(uint64,uint64)",
    values: [BigNumberish, BigNumberish]
  ): string;
  encodeFunctionData(
    functionFragment: "setTargetAdminDelay",
    values: [AddressLike, BigNumberish]
  ): string;
  encodeFunctionData(
    functionFragment: "setTargetAdminDelay(address,uint32)",
    values: [AddressLike, BigNumberish]
  ): string;
  encodeFunctionData(
    functionFragment: "setTargetClosed",
    values: [AddressLike, boolean]
  ): string;
  encodeFunctionData(
    functionFragment: "setTargetClosed(address,bool)",
    values: [AddressLike, boolean]
  ): string;
  encodeFunctionData(
    functionFragment: "setTargetFunctionRole",
    values: [AddressLike, BytesLike[], BigNumberish]
  ): string;
  encodeFunctionData(
    functionFragment: "setTargetFunctionRole(address,bytes4[],uint64)",
    values: [AddressLike, BytesLike[], BigNumberish]
  ): string;
  encodeFunctionData(
    functionFragment: "updateAuthority",
    values: [AddressLike, AddressLike]
  ): string;
  encodeFunctionData(
    functionFragment: "updateAuthority(address,address)",
    values: [AddressLike, AddressLike]
  ): string;

  decodeFunctionResult(functionFragment: "ADMIN_ROLE", data: BytesLike): Result;
  decodeFunctionResult(
    functionFragment: "ADMIN_ROLE()",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "PUBLIC_ROLE",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "PUBLIC_ROLE()",
    data: BytesLike
  ): Result;
  decodeFunctionResult(functionFragment: "canCall", data: BytesLike): Result;
  decodeFunctionResult(
    functionFragment: "canCall(address,address,bytes4)",
    data: BytesLike
  ): Result;
  decodeFunctionResult(functionFragment: "cancel", data: BytesLike): Result;
  decodeFunctionResult(
    functionFragment: "cancel(address,address,bytes)",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "consumeScheduledOp",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "consumeScheduledOp(address,bytes)",
    data: BytesLike
  ): Result;
  decodeFunctionResult(functionFragment: "execute", data: BytesLike): Result;
  decodeFunctionResult(
    functionFragment: "execute(address,bytes)",
    data: BytesLike
  ): Result;
  decodeFunctionResult(functionFragment: "expiration", data: BytesLike): Result;
  decodeFunctionResult(
    functionFragment: "expiration()",
    data: BytesLike
  ): Result;
  decodeFunctionResult(functionFragment: "getAccess", data: BytesLike): Result;
  decodeFunctionResult(
    functionFragment: "getAccess(uint64,address)",
    data: BytesLike
  ): Result;
  decodeFunctionResult(functionFragment: "getNonce", data: BytesLike): Result;
  decodeFunctionResult(
    functionFragment: "getNonce(bytes32)",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "getRoleAdmin",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "getRoleAdmin(uint64)",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "getRoleGrantDelay",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "getRoleGrantDelay(uint64)",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "getRoleGuardian",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "getRoleGuardian(uint64)",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "getSchedule",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "getSchedule(bytes32)",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "getTargetAdminDelay",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "getTargetAdminDelay(address)",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "getTargetFunctionRole",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "getTargetFunctionRole(address,bytes4)",
    data: BytesLike
  ): Result;
  decodeFunctionResult(functionFragment: "grantRole", data: BytesLike): Result;
  decodeFunctionResult(
    functionFragment: "grantRole(uint64,address,uint32)",
    data: BytesLike
  ): Result;
  decodeFunctionResult(functionFragment: "hasRole", data: BytesLike): Result;
  decodeFunctionResult(
    functionFragment: "hasRole(uint64,address)",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "hashOperation",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "hashOperation(address,address,bytes)",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "isTargetClosed",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "isTargetClosed(address)",
    data: BytesLike
  ): Result;
  decodeFunctionResult(functionFragment: "labelRole", data: BytesLike): Result;
  decodeFunctionResult(
    functionFragment: "labelRole(uint64,string)",
    data: BytesLike
  ): Result;
  decodeFunctionResult(functionFragment: "minSetback", data: BytesLike): Result;
  decodeFunctionResult(
    functionFragment: "minSetback()",
    data: BytesLike
  ): Result;
  decodeFunctionResult(functionFragment: "multicall", data: BytesLike): Result;
  decodeFunctionResult(
    functionFragment: "multicall(bytes[])",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "renounceRole",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "renounceRole(uint64,address)",
    data: BytesLike
  ): Result;
  decodeFunctionResult(functionFragment: "revokeRole", data: BytesLike): Result;
  decodeFunctionResult(
    functionFragment: "revokeRole(uint64,address)",
    data: BytesLike
  ): Result;
  decodeFunctionResult(functionFragment: "schedule", data: BytesLike): Result;
  decodeFunctionResult(
    functionFragment: "schedule(address,bytes,uint48)",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "setGrantDelay",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "setGrantDelay(uint64,uint32)",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "setRoleAdmin",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "setRoleAdmin(uint64,uint64)",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "setRoleGuardian",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "setRoleGuardian(uint64,uint64)",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "setTargetAdminDelay",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "setTargetAdminDelay(address,uint32)",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "setTargetClosed",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "setTargetClosed(address,bool)",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "setTargetFunctionRole",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "setTargetFunctionRole(address,bytes4[],uint64)",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "updateAuthority",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "updateAuthority(address,address)",
    data: BytesLike
  ): Result;
}

export namespace OperationCanceledEvent {
  export type InputTuple = [operationId: BytesLike, nonce: BigNumberish];
  export type OutputTuple = [operationId: string, nonce: bigint];
  export interface OutputObject {
    operationId: string;
    nonce: bigint;
  }
  export type Event = TypedContractEvent<InputTuple, OutputTuple, OutputObject>;
  export type Filter = TypedDeferredTopicFilter<Event>;
  export type Log = TypedEventLog<Event>;
  export type LogDescription = TypedLogDescription<Event>;
}

export namespace OperationExecutedEvent {
  export type InputTuple = [operationId: BytesLike, nonce: BigNumberish];
  export type OutputTuple = [operationId: string, nonce: bigint];
  export interface OutputObject {
    operationId: string;
    nonce: bigint;
  }
  export type Event = TypedContractEvent<InputTuple, OutputTuple, OutputObject>;
  export type Filter = TypedDeferredTopicFilter<Event>;
  export type Log = TypedEventLog<Event>;
  export type LogDescription = TypedLogDescription<Event>;
}

export namespace OperationScheduledEvent {
  export type InputTuple = [
    operationId: BytesLike,
    nonce: BigNumberish,
    schedule: BigNumberish,
    caller: AddressLike,
    target: AddressLike,
    data: BytesLike
  ];
  export type OutputTuple = [
    operationId: string,
    nonce: bigint,
    schedule: bigint,
    caller: string,
    target: string,
    data: string
  ];
  export interface OutputObject {
    operationId: string;
    nonce: bigint;
    schedule: bigint;
    caller: string;
    target: string;
    data: string;
  }
  export type Event = TypedContractEvent<InputTuple, OutputTuple, OutputObject>;
  export type Filter = TypedDeferredTopicFilter<Event>;
  export type Log = TypedEventLog<Event>;
  export type LogDescription = TypedLogDescription<Event>;
}

export namespace RoleAdminChangedEvent {
  export type InputTuple = [roleId: BigNumberish, admin: BigNumberish];
  export type OutputTuple = [roleId: bigint, admin: bigint];
  export interface OutputObject {
    roleId: bigint;
    admin: bigint;
  }
  export type Event = TypedContractEvent<InputTuple, OutputTuple, OutputObject>;
  export type Filter = TypedDeferredTopicFilter<Event>;
  export type Log = TypedEventLog<Event>;
  export type LogDescription = TypedLogDescription<Event>;
}

export namespace RoleGrantDelayChangedEvent {
  export type InputTuple = [
    roleId: BigNumberish,
    delay: BigNumberish,
    since: BigNumberish
  ];
  export type OutputTuple = [roleId: bigint, delay: bigint, since: bigint];
  export interface OutputObject {
    roleId: bigint;
    delay: bigint;
    since: bigint;
  }
  export type Event = TypedContractEvent<InputTuple, OutputTuple, OutputObject>;
  export type Filter = TypedDeferredTopicFilter<Event>;
  export type Log = TypedEventLog<Event>;
  export type LogDescription = TypedLogDescription<Event>;
}

export namespace RoleGrantedEvent {
  export type InputTuple = [
    roleId: BigNumberish,
    account: AddressLike,
    delay: BigNumberish,
    since: BigNumberish,
    newMember: boolean
  ];
  export type OutputTuple = [
    roleId: bigint,
    account: string,
    delay: bigint,
    since: bigint,
    newMember: boolean
  ];
  export interface OutputObject {
    roleId: bigint;
    account: string;
    delay: bigint;
    since: bigint;
    newMember: boolean;
  }
  export type Event = TypedContractEvent<InputTuple, OutputTuple, OutputObject>;
  export type Filter = TypedDeferredTopicFilter<Event>;
  export type Log = TypedEventLog<Event>;
  export type LogDescription = TypedLogDescription<Event>;
}

export namespace RoleGuardianChangedEvent {
  export type InputTuple = [roleId: BigNumberish, guardian: BigNumberish];
  export type OutputTuple = [roleId: bigint, guardian: bigint];
  export interface OutputObject {
    roleId: bigint;
    guardian: bigint;
  }
  export type Event = TypedContractEvent<InputTuple, OutputTuple, OutputObject>;
  export type Filter = TypedDeferredTopicFilter<Event>;
  export type Log = TypedEventLog<Event>;
  export type LogDescription = TypedLogDescription<Event>;
}

export namespace RoleLabelEvent {
  export type InputTuple = [roleId: BigNumberish, label: string];
  export type OutputTuple = [roleId: bigint, label: string];
  export interface OutputObject {
    roleId: bigint;
    label: string;
  }
  export type Event = TypedContractEvent<InputTuple, OutputTuple, OutputObject>;
  export type Filter = TypedDeferredTopicFilter<Event>;
  export type Log = TypedEventLog<Event>;
  export type LogDescription = TypedLogDescription<Event>;
}

export namespace RoleRevokedEvent {
  export type InputTuple = [roleId: BigNumberish, account: AddressLike];
  export type OutputTuple = [roleId: bigint, account: string];
  export interface OutputObject {
    roleId: bigint;
    account: string;
  }
  export type Event = TypedContractEvent<InputTuple, OutputTuple, OutputObject>;
  export type Filter = TypedDeferredTopicFilter<Event>;
  export type Log = TypedEventLog<Event>;
  export type LogDescription = TypedLogDescription<Event>;
}

export namespace TargetAdminDelayUpdatedEvent {
  export type InputTuple = [
    target: AddressLike,
    delay: BigNumberish,
    since: BigNumberish
  ];
  export type OutputTuple = [target: string, delay: bigint, since: bigint];
  export interface OutputObject {
    target: string;
    delay: bigint;
    since: bigint;
  }
  export type Event = TypedContractEvent<InputTuple, OutputTuple, OutputObject>;
  export type Filter = TypedDeferredTopicFilter<Event>;
  export type Log = TypedEventLog<Event>;
  export type LogDescription = TypedLogDescription<Event>;
}

export namespace TargetClosedEvent {
  export type InputTuple = [target: AddressLike, closed: boolean];
  export type OutputTuple = [target: string, closed: boolean];
  export interface OutputObject {
    target: string;
    closed: boolean;
  }
  export type Event = TypedContractEvent<InputTuple, OutputTuple, OutputObject>;
  export type Filter = TypedDeferredTopicFilter<Event>;
  export type Log = TypedEventLog<Event>;
  export type LogDescription = TypedLogDescription<Event>;
}

export namespace TargetFunctionRoleUpdatedEvent {
  export type InputTuple = [
    target: AddressLike,
    selector: BytesLike,
    roleId: BigNumberish
  ];
  export type OutputTuple = [target: string, selector: string, roleId: bigint];
  export interface OutputObject {
    target: string;
    selector: string;
    roleId: bigint;
  }
  export type Event = TypedContractEvent<InputTuple, OutputTuple, OutputObject>;
  export type Filter = TypedDeferredTopicFilter<Event>;
  export type Log = TypedEventLog<Event>;
  export type LogDescription = TypedLogDescription<Event>;
}

export interface AccessManager extends BaseContract {
  connect(runner?: ContractRunner | null): AccessManager;
  waitForDeployment(): Promise<this>;

  interface: AccessManagerInterface;

  queryFilter<TCEvent extends TypedContractEvent>(
    event: TCEvent,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TypedEventLog<TCEvent>>>;
  queryFilter<TCEvent extends TypedContractEvent>(
    filter: TypedDeferredTopicFilter<TCEvent>,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TypedEventLog<TCEvent>>>;

  on<TCEvent extends TypedContractEvent>(
    event: TCEvent,
    listener: TypedListener<TCEvent>
  ): Promise<this>;
  on<TCEvent extends TypedContractEvent>(
    filter: TypedDeferredTopicFilter<TCEvent>,
    listener: TypedListener<TCEvent>
  ): Promise<this>;

  once<TCEvent extends TypedContractEvent>(
    event: TCEvent,
    listener: TypedListener<TCEvent>
  ): Promise<this>;
  once<TCEvent extends TypedContractEvent>(
    filter: TypedDeferredTopicFilter<TCEvent>,
    listener: TypedListener<TCEvent>
  ): Promise<this>;

  listeners<TCEvent extends TypedContractEvent>(
    event: TCEvent
  ): Promise<Array<TypedListener<TCEvent>>>;
  listeners(eventName?: string): Promise<Array<Listener>>;
  removeAllListeners<TCEvent extends TypedContractEvent>(
    event?: TCEvent
  ): Promise<this>;

  ADMIN_ROLE: TypedContractMethod<[], [bigint], "view">;

  "ADMIN_ROLE()": TypedContractMethod<[], [bigint], "view">;

  PUBLIC_ROLE: TypedContractMethod<[], [bigint], "view">;

  "PUBLIC_ROLE()": TypedContractMethod<[], [bigint], "view">;

  canCall: TypedContractMethod<
    [caller: AddressLike, target: AddressLike, selector: BytesLike],
    [[boolean, bigint] & { immediate: boolean; delay: bigint }],
    "view"
  >;

  "canCall(address,address,bytes4)": TypedContractMethod<
    [caller: AddressLike, target: AddressLike, selector: BytesLike],
    [[boolean, bigint] & { immediate: boolean; delay: bigint }],
    "view"
  >;

  cancel: TypedContractMethod<
    [caller: AddressLike, target: AddressLike, data: BytesLike],
    [bigint],
    "nonpayable"
  >;

  "cancel(address,address,bytes)": TypedContractMethod<
    [caller: AddressLike, target: AddressLike, data: BytesLike],
    [bigint],
    "nonpayable"
  >;

  consumeScheduledOp: TypedContractMethod<
    [caller: AddressLike, data: BytesLike],
    [void],
    "nonpayable"
  >;

  "consumeScheduledOp(address,bytes)": TypedContractMethod<
    [caller: AddressLike, data: BytesLike],
    [void],
    "nonpayable"
  >;

  execute: TypedContractMethod<
    [target: AddressLike, data: BytesLike],
    [bigint],
    "payable"
  >;

  "execute(address,bytes)": TypedContractMethod<
    [target: AddressLike, data: BytesLike],
    [bigint],
    "payable"
  >;

  expiration: TypedContractMethod<[], [bigint], "view">;

  "expiration()": TypedContractMethod<[], [bigint], "view">;

  getAccess: TypedContractMethod<
    [roleId: BigNumberish, account: AddressLike],
    [
      [bigint, bigint, bigint, bigint] & {
        since: bigint;
        currentDelay: bigint;
        pendingDelay: bigint;
        effect: bigint;
      }
    ],
    "view"
  >;

  "getAccess(uint64,address)": TypedContractMethod<
    [roleId: BigNumberish, account: AddressLike],
    [
      [bigint, bigint, bigint, bigint] & {
        since: bigint;
        currentDelay: bigint;
        pendingDelay: bigint;
        effect: bigint;
      }
    ],
    "view"
  >;

  getNonce: TypedContractMethod<[id: BytesLike], [bigint], "view">;

  "getNonce(bytes32)": TypedContractMethod<[id: BytesLike], [bigint], "view">;

  getRoleAdmin: TypedContractMethod<[roleId: BigNumberish], [bigint], "view">;

  "getRoleAdmin(uint64)": TypedContractMethod<
    [roleId: BigNumberish],
    [bigint],
    "view"
  >;

  getRoleGrantDelay: TypedContractMethod<
    [roleId: BigNumberish],
    [bigint],
    "view"
  >;

  "getRoleGrantDelay(uint64)": TypedContractMethod<
    [roleId: BigNumberish],
    [bigint],
    "view"
  >;

  getRoleGuardian: TypedContractMethod<
    [roleId: BigNumberish],
    [bigint],
    "view"
  >;

  "getRoleGuardian(uint64)": TypedContractMethod<
    [roleId: BigNumberish],
    [bigint],
    "view"
  >;

  getSchedule: TypedContractMethod<[id: BytesLike], [bigint], "view">;

  "getSchedule(bytes32)": TypedContractMethod<
    [id: BytesLike],
    [bigint],
    "view"
  >;

  getTargetAdminDelay: TypedContractMethod<
    [target: AddressLike],
    [bigint],
    "view"
  >;

  "getTargetAdminDelay(address)": TypedContractMethod<
    [target: AddressLike],
    [bigint],
    "view"
  >;

  getTargetFunctionRole: TypedContractMethod<
    [target: AddressLike, selector: BytesLike],
    [bigint],
    "view"
  >;

  "getTargetFunctionRole(address,bytes4)": TypedContractMethod<
    [target: AddressLike, selector: BytesLike],
    [bigint],
    "view"
  >;

  grantRole: TypedContractMethod<
    [roleId: BigNumberish, account: AddressLike, executionDelay: BigNumberish],
    [void],
    "nonpayable"
  >;

  "grantRole(uint64,address,uint32)": TypedContractMethod<
    [roleId: BigNumberish, account: AddressLike, executionDelay: BigNumberish],
    [void],
    "nonpayable"
  >;

  hasRole: TypedContractMethod<
    [roleId: BigNumberish, account: AddressLike],
    [[boolean, bigint] & { isMember: boolean; executionDelay: bigint }],
    "view"
  >;

  "hasRole(uint64,address)": TypedContractMethod<
    [roleId: BigNumberish, account: AddressLike],
    [[boolean, bigint] & { isMember: boolean; executionDelay: bigint }],
    "view"
  >;

  hashOperation: TypedContractMethod<
    [caller: AddressLike, target: AddressLike, data: BytesLike],
    [string],
    "view"
  >;

  "hashOperation(address,address,bytes)": TypedContractMethod<
    [caller: AddressLike, target: AddressLike, data: BytesLike],
    [string],
    "view"
  >;

  isTargetClosed: TypedContractMethod<[target: AddressLike], [boolean], "view">;

  "isTargetClosed(address)": TypedContractMethod<
    [target: AddressLike],
    [boolean],
    "view"
  >;

  labelRole: TypedContractMethod<
    [roleId: BigNumberish, label: string],
    [void],
    "nonpayable"
  >;

  "labelRole(uint64,string)": TypedContractMethod<
    [roleId: BigNumberish, label: string],
    [void],
    "nonpayable"
  >;

  minSetback: TypedContractMethod<[], [bigint], "view">;

  "minSetback()": TypedContractMethod<[], [bigint], "view">;

  multicall: TypedContractMethod<[data: BytesLike[]], [string[]], "nonpayable">;

  "multicall(bytes[])": TypedContractMethod<
    [data: BytesLike[]],
    [string[]],
    "nonpayable"
  >;

  renounceRole: TypedContractMethod<
    [roleId: BigNumberish, callerConfirmation: AddressLike],
    [void],
    "nonpayable"
  >;

  "renounceRole(uint64,address)": TypedContractMethod<
    [roleId: BigNumberish, callerConfirmation: AddressLike],
    [void],
    "nonpayable"
  >;

  revokeRole: TypedContractMethod<
    [roleId: BigNumberish, account: AddressLike],
    [void],
    "nonpayable"
  >;

  "revokeRole(uint64,address)": TypedContractMethod<
    [roleId: BigNumberish, account: AddressLike],
    [void],
    "nonpayable"
  >;

  schedule: TypedContractMethod<
    [target: AddressLike, data: BytesLike, when: BigNumberish],
    [[string, bigint] & { operationId: string; nonce: bigint }],
    "nonpayable"
  >;

  "schedule(address,bytes,uint48)": TypedContractMethod<
    [target: AddressLike, data: BytesLike, when: BigNumberish],
    [[string, bigint] & { operationId: string; nonce: bigint }],
    "nonpayable"
  >;

  setGrantDelay: TypedContractMethod<
    [roleId: BigNumberish, newDelay: BigNumberish],
    [void],
    "nonpayable"
  >;

  "setGrantDelay(uint64,uint32)": TypedContractMethod<
    [roleId: BigNumberish, newDelay: BigNumberish],
    [void],
    "nonpayable"
  >;

  setRoleAdmin: TypedContractMethod<
    [roleId: BigNumberish, admin: BigNumberish],
    [void],
    "nonpayable"
  >;

  "setRoleAdmin(uint64,uint64)": TypedContractMethod<
    [roleId: BigNumberish, admin: BigNumberish],
    [void],
    "nonpayable"
  >;

  setRoleGuardian: TypedContractMethod<
    [roleId: BigNumberish, guardian: BigNumberish],
    [void],
    "nonpayable"
  >;

  "setRoleGuardian(uint64,uint64)": TypedContractMethod<
    [roleId: BigNumberish, guardian: BigNumberish],
    [void],
    "nonpayable"
  >;

  setTargetAdminDelay: TypedContractMethod<
    [target: AddressLike, newDelay: BigNumberish],
    [void],
    "nonpayable"
  >;

  "setTargetAdminDelay(address,uint32)": TypedContractMethod<
    [target: AddressLike, newDelay: BigNumberish],
    [void],
    "nonpayable"
  >;

  setTargetClosed: TypedContractMethod<
    [target: AddressLike, closed: boolean],
    [void],
    "nonpayable"
  >;

  "setTargetClosed(address,bool)": TypedContractMethod<
    [target: AddressLike, closed: boolean],
    [void],
    "nonpayable"
  >;

  setTargetFunctionRole: TypedContractMethod<
    [target: AddressLike, selectors: BytesLike[], roleId: BigNumberish],
    [void],
    "nonpayable"
  >;

  "setTargetFunctionRole(address,bytes4[],uint64)": TypedContractMethod<
    [target: AddressLike, selectors: BytesLike[], roleId: BigNumberish],
    [void],
    "nonpayable"
  >;

  updateAuthority: TypedContractMethod<
    [target: AddressLike, newAuthority: AddressLike],
    [void],
    "nonpayable"
  >;

  "updateAuthority(address,address)": TypedContractMethod<
    [target: AddressLike, newAuthority: AddressLike],
    [void],
    "nonpayable"
  >;

  getFunction<T extends ContractMethod = ContractMethod>(
    key: string | FunctionFragment
  ): T;

  getFunction(
    nameOrSignature: "ADMIN_ROLE"
  ): TypedContractMethod<[], [bigint], "view">;
  getFunction(
    nameOrSignature: "ADMIN_ROLE()"
  ): TypedContractMethod<[], [bigint], "view">;
  getFunction(
    nameOrSignature: "PUBLIC_ROLE"
  ): TypedContractMethod<[], [bigint], "view">;
  getFunction(
    nameOrSignature: "PUBLIC_ROLE()"
  ): TypedContractMethod<[], [bigint], "view">;
  getFunction(
    nameOrSignature: "canCall"
  ): TypedContractMethod<
    [caller: AddressLike, target: AddressLike, selector: BytesLike],
    [[boolean, bigint] & { immediate: boolean; delay: bigint }],
    "view"
  >;
  getFunction(
    nameOrSignature: "canCall(address,address,bytes4)"
  ): TypedContractMethod<
    [caller: AddressLike, target: AddressLike, selector: BytesLike],
    [[boolean, bigint] & { immediate: boolean; delay: bigint }],
    "view"
  >;
  getFunction(
    nameOrSignature: "cancel"
  ): TypedContractMethod<
    [caller: AddressLike, target: AddressLike, data: BytesLike],
    [bigint],
    "nonpayable"
  >;
  getFunction(
    nameOrSignature: "cancel(address,address,bytes)"
  ): TypedContractMethod<
    [caller: AddressLike, target: AddressLike, data: BytesLike],
    [bigint],
    "nonpayable"
  >;
  getFunction(
    nameOrSignature: "consumeScheduledOp"
  ): TypedContractMethod<
    [caller: AddressLike, data: BytesLike],
    [void],
    "nonpayable"
  >;
  getFunction(
    nameOrSignature: "consumeScheduledOp(address,bytes)"
  ): TypedContractMethod<
    [caller: AddressLike, data: BytesLike],
    [void],
    "nonpayable"
  >;
  getFunction(
    nameOrSignature: "execute"
  ): TypedContractMethod<
    [target: AddressLike, data: BytesLike],
    [bigint],
    "payable"
  >;
  getFunction(
    nameOrSignature: "execute(address,bytes)"
  ): TypedContractMethod<
    [target: AddressLike, data: BytesLike],
    [bigint],
    "payable"
  >;
  getFunction(
    nameOrSignature: "expiration"
  ): TypedContractMethod<[], [bigint], "view">;
  getFunction(
    nameOrSignature: "expiration()"
  ): TypedContractMethod<[], [bigint], "view">;
  getFunction(
    nameOrSignature: "getAccess"
  ): TypedContractMethod<
    [roleId: BigNumberish, account: AddressLike],
    [
      [bigint, bigint, bigint, bigint] & {
        since: bigint;
        currentDelay: bigint;
        pendingDelay: bigint;
        effect: bigint;
      }
    ],
    "view"
  >;
  getFunction(
    nameOrSignature: "getAccess(uint64,address)"
  ): TypedContractMethod<
    [roleId: BigNumberish, account: AddressLike],
    [
      [bigint, bigint, bigint, bigint] & {
        since: bigint;
        currentDelay: bigint;
        pendingDelay: bigint;
        effect: bigint;
      }
    ],
    "view"
  >;
  getFunction(
    nameOrSignature: "getNonce"
  ): TypedContractMethod<[id: BytesLike], [bigint], "view">;
  getFunction(
    nameOrSignature: "getNonce(bytes32)"
  ): TypedContractMethod<[id: BytesLike], [bigint], "view">;
  getFunction(
    nameOrSignature: "getRoleAdmin"
  ): TypedContractMethod<[roleId: BigNumberish], [bigint], "view">;
  getFunction(
    nameOrSignature: "getRoleAdmin(uint64)"
  ): TypedContractMethod<[roleId: BigNumberish], [bigint], "view">;
  getFunction(
    nameOrSignature: "getRoleGrantDelay"
  ): TypedContractMethod<[roleId: BigNumberish], [bigint], "view">;
  getFunction(
    nameOrSignature: "getRoleGrantDelay(uint64)"
  ): TypedContractMethod<[roleId: BigNumberish], [bigint], "view">;
  getFunction(
    nameOrSignature: "getRoleGuardian"
  ): TypedContractMethod<[roleId: BigNumberish], [bigint], "view">;
  getFunction(
    nameOrSignature: "getRoleGuardian(uint64)"
  ): TypedContractMethod<[roleId: BigNumberish], [bigint], "view">;
  getFunction(
    nameOrSignature: "getSchedule"
  ): TypedContractMethod<[id: BytesLike], [bigint], "view">;
  getFunction(
    nameOrSignature: "getSchedule(bytes32)"
  ): TypedContractMethod<[id: BytesLike], [bigint], "view">;
  getFunction(
    nameOrSignature: "getTargetAdminDelay"
  ): TypedContractMethod<[target: AddressLike], [bigint], "view">;
  getFunction(
    nameOrSignature: "getTargetAdminDelay(address)"
  ): TypedContractMethod<[target: AddressLike], [bigint], "view">;
  getFunction(
    nameOrSignature: "getTargetFunctionRole"
  ): TypedContractMethod<
    [target: AddressLike, selector: BytesLike],
    [bigint],
    "view"
  >;
  getFunction(
    nameOrSignature: "getTargetFunctionRole(address,bytes4)"
  ): TypedContractMethod<
    [target: AddressLike, selector: BytesLike],
    [bigint],
    "view"
  >;
  getFunction(
    nameOrSignature: "grantRole"
  ): TypedContractMethod<
    [roleId: BigNumberish, account: AddressLike, executionDelay: BigNumberish],
    [void],
    "nonpayable"
  >;
  getFunction(
    nameOrSignature: "grantRole(uint64,address,uint32)"
  ): TypedContractMethod<
    [roleId: BigNumberish, account: AddressLike, executionDelay: BigNumberish],
    [void],
    "nonpayable"
  >;
  getFunction(
    nameOrSignature: "hasRole"
  ): TypedContractMethod<
    [roleId: BigNumberish, account: AddressLike],
    [[boolean, bigint] & { isMember: boolean; executionDelay: bigint }],
    "view"
  >;
  getFunction(
    nameOrSignature: "hasRole(uint64,address)"
  ): TypedContractMethod<
    [roleId: BigNumberish, account: AddressLike],
    [[boolean, bigint] & { isMember: boolean; executionDelay: bigint }],
    "view"
  >;
  getFunction(
    nameOrSignature: "hashOperation"
  ): TypedContractMethod<
    [caller: AddressLike, target: AddressLike, data: BytesLike],
    [string],
    "view"
  >;
  getFunction(
    nameOrSignature: "hashOperation(address,address,bytes)"
  ): TypedContractMethod<
    [caller: AddressLike, target: AddressLike, data: BytesLike],
    [string],
    "view"
  >;
  getFunction(
    nameOrSignature: "isTargetClosed"
  ): TypedContractMethod<[target: AddressLike], [boolean], "view">;
  getFunction(
    nameOrSignature: "isTargetClosed(address)"
  ): TypedContractMethod<[target: AddressLike], [boolean], "view">;
  getFunction(
    nameOrSignature: "labelRole"
  ): TypedContractMethod<
    [roleId: BigNumberish, label: string],
    [void],
    "nonpayable"
  >;
  getFunction(
    nameOrSignature: "labelRole(uint64,string)"
  ): TypedContractMethod<
    [roleId: BigNumberish, label: string],
    [void],
    "nonpayable"
  >;
  getFunction(
    nameOrSignature: "minSetback"
  ): TypedContractMethod<[], [bigint], "view">;
  getFunction(
    nameOrSignature: "minSetback()"
  ): TypedContractMethod<[], [bigint], "view">;
  getFunction(
    nameOrSignature: "multicall"
  ): TypedContractMethod<[data: BytesLike[]], [string[]], "nonpayable">;
  getFunction(
    nameOrSignature: "multicall(bytes[])"
  ): TypedContractMethod<[data: BytesLike[]], [string[]], "nonpayable">;
  getFunction(
    nameOrSignature: "renounceRole"
  ): TypedContractMethod<
    [roleId: BigNumberish, callerConfirmation: AddressLike],
    [void],
    "nonpayable"
  >;
  getFunction(
    nameOrSignature: "renounceRole(uint64,address)"
  ): TypedContractMethod<
    [roleId: BigNumberish, callerConfirmation: AddressLike],
    [void],
    "nonpayable"
  >;
  getFunction(
    nameOrSignature: "revokeRole"
  ): TypedContractMethod<
    [roleId: BigNumberish, account: AddressLike],
    [void],
    "nonpayable"
  >;
  getFunction(
    nameOrSignature: "revokeRole(uint64,address)"
  ): TypedContractMethod<
    [roleId: BigNumberish, account: AddressLike],
    [void],
    "nonpayable"
  >;
  getFunction(
    nameOrSignature: "schedule"
  ): TypedContractMethod<
    [target: AddressLike, data: BytesLike, when: BigNumberish],
    [[string, bigint] & { operationId: string; nonce: bigint }],
    "nonpayable"
  >;
  getFunction(
    nameOrSignature: "schedule(address,bytes,uint48)"
  ): TypedContractMethod<
    [target: AddressLike, data: BytesLike, when: BigNumberish],
    [[string, bigint] & { operationId: string; nonce: bigint }],
    "nonpayable"
  >;
  getFunction(
    nameOrSignature: "setGrantDelay"
  ): TypedContractMethod<
    [roleId: BigNumberish, newDelay: BigNumberish],
    [void],
    "nonpayable"
  >;
  getFunction(
    nameOrSignature: "setGrantDelay(uint64,uint32)"
  ): TypedContractMethod<
    [roleId: BigNumberish, newDelay: BigNumberish],
    [void],
    "nonpayable"
  >;
  getFunction(
    nameOrSignature: "setRoleAdmin"
  ): TypedContractMethod<
    [roleId: BigNumberish, admin: BigNumberish],
    [void],
    "nonpayable"
  >;
  getFunction(
    nameOrSignature: "setRoleAdmin(uint64,uint64)"
  ): TypedContractMethod<
    [roleId: BigNumberish, admin: BigNumberish],
    [void],
    "nonpayable"
  >;
  getFunction(
    nameOrSignature: "setRoleGuardian"
  ): TypedContractMethod<
    [roleId: BigNumberish, guardian: BigNumberish],
    [void],
    "nonpayable"
  >;
  getFunction(
    nameOrSignature: "setRoleGuardian(uint64,uint64)"
  ): TypedContractMethod<
    [roleId: BigNumberish, guardian: BigNumberish],
    [void],
    "nonpayable"
  >;
  getFunction(
    nameOrSignature: "setTargetAdminDelay"
  ): TypedContractMethod<
    [target: AddressLike, newDelay: BigNumberish],
    [void],
    "nonpayable"
  >;
  getFunction(
    nameOrSignature: "setTargetAdminDelay(address,uint32)"
  ): TypedContractMethod<
    [target: AddressLike, newDelay: BigNumberish],
    [void],
    "nonpayable"
  >;
  getFunction(
    nameOrSignature: "setTargetClosed"
  ): TypedContractMethod<
    [target: AddressLike, closed: boolean],
    [void],
    "nonpayable"
  >;
  getFunction(
    nameOrSignature: "setTargetClosed(address,bool)"
  ): TypedContractMethod<
    [target: AddressLike, closed: boolean],
    [void],
    "nonpayable"
  >;
  getFunction(
    nameOrSignature: "setTargetFunctionRole"
  ): TypedContractMethod<
    [target: AddressLike, selectors: BytesLike[], roleId: BigNumberish],
    [void],
    "nonpayable"
  >;
  getFunction(
    nameOrSignature: "setTargetFunctionRole(address,bytes4[],uint64)"
  ): TypedContractMethod<
    [target: AddressLike, selectors: BytesLike[], roleId: BigNumberish],
    [void],
    "nonpayable"
  >;
  getFunction(
    nameOrSignature: "updateAuthority"
  ): TypedContractMethod<
    [target: AddressLike, newAuthority: AddressLike],
    [void],
    "nonpayable"
  >;
  getFunction(
    nameOrSignature: "updateAuthority(address,address)"
  ): TypedContractMethod<
    [target: AddressLike, newAuthority: AddressLike],
    [void],
    "nonpayable"
  >;

  getEvent(
    key: "OperationCanceled"
  ): TypedContractEvent<
    OperationCanceledEvent.InputTuple,
    OperationCanceledEvent.OutputTuple,
    OperationCanceledEvent.OutputObject
  >;
  getEvent(
    key: "OperationCanceled(bytes32,uint32)"
  ): TypedContractEvent<
    OperationCanceled_bytes32_uint32_Event.InputTuple,
    OperationCanceled_bytes32_uint32_Event.OutputTuple,
    OperationCanceled_bytes32_uint32_Event.OutputObject
  >;
  getEvent(
    key: "OperationExecuted"
  ): TypedContractEvent<
    OperationExecutedEvent.InputTuple,
    OperationExecutedEvent.OutputTuple,
    OperationExecutedEvent.OutputObject
  >;
  getEvent(
    key: "OperationExecuted(bytes32,uint32)"
  ): TypedContractEvent<
    OperationExecuted_bytes32_uint32_Event.InputTuple,
    OperationExecuted_bytes32_uint32_Event.OutputTuple,
    OperationExecuted_bytes32_uint32_Event.OutputObject
  >;
  getEvent(
    key: "OperationScheduled"
  ): TypedContractEvent<
    OperationScheduledEvent.InputTuple,
    OperationScheduledEvent.OutputTuple,
    OperationScheduledEvent.OutputObject
  >;
  getEvent(
    key: "OperationScheduled(bytes32,uint32,uint48,address,address,bytes)"
  ): TypedContractEvent<
    OperationScheduled_bytes32_uint32_uint48_address_address_bytes_Event.InputTuple,
    OperationScheduled_bytes32_uint32_uint48_address_address_bytes_Event.OutputTuple,
    OperationScheduled_bytes32_uint32_uint48_address_address_bytes_Event.OutputObject
  >;
  getEvent(
    key: "RoleAdminChanged"
  ): TypedContractEvent<
    RoleAdminChangedEvent.InputTuple,
    RoleAdminChangedEvent.OutputTuple,
    RoleAdminChangedEvent.OutputObject
  >;
  getEvent(
    key: "RoleAdminChanged(uint64,uint64)"
  ): TypedContractEvent<
    RoleAdminChanged_uint64_uint64_Event.InputTuple,
    RoleAdminChanged_uint64_uint64_Event.OutputTuple,
    RoleAdminChanged_uint64_uint64_Event.OutputObject
  >;
  getEvent(
    key: "RoleGrantDelayChanged"
  ): TypedContractEvent<
    RoleGrantDelayChangedEvent.InputTuple,
    RoleGrantDelayChangedEvent.OutputTuple,
    RoleGrantDelayChangedEvent.OutputObject
  >;
  getEvent(
    key: "RoleGrantDelayChanged(uint64,uint32,uint48)"
  ): TypedContractEvent<
    RoleGrantDelayChanged_uint64_uint32_uint48_Event.InputTuple,
    RoleGrantDelayChanged_uint64_uint32_uint48_Event.OutputTuple,
    RoleGrantDelayChanged_uint64_uint32_uint48_Event.OutputObject
  >;
  getEvent(
    key: "RoleGranted"
  ): TypedContractEvent<
    RoleGrantedEvent.InputTuple,
    RoleGrantedEvent.OutputTuple,
    RoleGrantedEvent.OutputObject
  >;
  getEvent(
    key: "RoleGranted(uint64,address,uint32,uint48,bool)"
  ): TypedContractEvent<
    RoleGranted_uint64_address_uint32_uint48_bool_Event.InputTuple,
    RoleGranted_uint64_address_uint32_uint48_bool_Event.OutputTuple,
    RoleGranted_uint64_address_uint32_uint48_bool_Event.OutputObject
  >;
  getEvent(
    key: "RoleGuardianChanged"
  ): TypedContractEvent<
    RoleGuardianChangedEvent.InputTuple,
    RoleGuardianChangedEvent.OutputTuple,
    RoleGuardianChangedEvent.OutputObject
  >;
  getEvent(
    key: "RoleGuardianChanged(uint64,uint64)"
  ): TypedContractEvent<
    RoleGuardianChanged_uint64_uint64_Event.InputTuple,
    RoleGuardianChanged_uint64_uint64_Event.OutputTuple,
    RoleGuardianChanged_uint64_uint64_Event.OutputObject
  >;
  getEvent(
    key: "RoleLabel"
  ): TypedContractEvent<
    RoleLabelEvent.InputTuple,
    RoleLabelEvent.OutputTuple,
    RoleLabelEvent.OutputObject
  >;
  getEvent(
    key: "RoleLabel(uint64,string)"
  ): TypedContractEvent<
    RoleLabel_uint64_string_Event.InputTuple,
    RoleLabel_uint64_string_Event.OutputTuple,
    RoleLabel_uint64_string_Event.OutputObject
  >;
  getEvent(
    key: "RoleRevoked"
  ): TypedContractEvent<
    RoleRevokedEvent.InputTuple,
    RoleRevokedEvent.OutputTuple,
    RoleRevokedEvent.OutputObject
  >;
  getEvent(
    key: "RoleRevoked(uint64,address)"
  ): TypedContractEvent<
    RoleRevoked_uint64_address_Event.InputTuple,
    RoleRevoked_uint64_address_Event.OutputTuple,
    RoleRevoked_uint64_address_Event.OutputObject
  >;
  getEvent(
    key: "TargetAdminDelayUpdated"
  ): TypedContractEvent<
    TargetAdminDelayUpdatedEvent.InputTuple,
    TargetAdminDelayUpdatedEvent.OutputTuple,
    TargetAdminDelayUpdatedEvent.OutputObject
  >;
  getEvent(
    key: "TargetAdminDelayUpdated(address,uint32,uint48)"
  ): TypedContractEvent<
    TargetAdminDelayUpdated_address_uint32_uint48_Event.InputTuple,
    TargetAdminDelayUpdated_address_uint32_uint48_Event.OutputTuple,
    TargetAdminDelayUpdated_address_uint32_uint48_Event.OutputObject
  >;
  getEvent(
    key: "TargetClosed"
  ): TypedContractEvent<
    TargetClosedEvent.InputTuple,
    TargetClosedEvent.OutputTuple,
    TargetClosedEvent.OutputObject
  >;
  getEvent(
    key: "TargetClosed(address,bool)"
  ): TypedContractEvent<
    TargetClosed_address_bool_Event.InputTuple,
    TargetClosed_address_bool_Event.OutputTuple,
    TargetClosed_address_bool_Event.OutputObject
  >;
  getEvent(
    key: "TargetFunctionRoleUpdated"
  ): TypedContractEvent<
    TargetFunctionRoleUpdatedEvent.InputTuple,
    TargetFunctionRoleUpdatedEvent.OutputTuple,
    TargetFunctionRoleUpdatedEvent.OutputObject
  >;
  getEvent(
    key: "TargetFunctionRoleUpdated(address,bytes4,uint64)"
  ): TypedContractEvent<
    TargetFunctionRoleUpdated_address_bytes4_uint64_Event.InputTuple,
    TargetFunctionRoleUpdated_address_bytes4_uint64_Event.OutputTuple,
    TargetFunctionRoleUpdated_address_bytes4_uint64_Event.OutputObject
  >;

  filters: {
    "OperationCanceled(bytes32,uint32)": TypedContractEvent<
      OperationCanceledEvent.InputTuple,
      OperationCanceledEvent.OutputTuple,
      OperationCanceledEvent.OutputObject
    >;
    OperationCanceled: TypedContractEvent<
      OperationCanceledEvent.InputTuple,
      OperationCanceledEvent.OutputTuple,
      OperationCanceledEvent.OutputObject
    >;

    "OperationExecuted(bytes32,uint32)": TypedContractEvent<
      OperationExecutedEvent.InputTuple,
      OperationExecutedEvent.OutputTuple,
      OperationExecutedEvent.OutputObject
    >;
    OperationExecuted: TypedContractEvent<
      OperationExecutedEvent.InputTuple,
      OperationExecutedEvent.OutputTuple,
      OperationExecutedEvent.OutputObject
    >;

    "OperationScheduled(bytes32,uint32,uint48,address,address,bytes)": TypedContractEvent<
      OperationScheduledEvent.InputTuple,
      OperationScheduledEvent.OutputTuple,
      OperationScheduledEvent.OutputObject
    >;
    OperationScheduled: TypedContractEvent<
      OperationScheduledEvent.InputTuple,
      OperationScheduledEvent.OutputTuple,
      OperationScheduledEvent.OutputObject
    >;

    "RoleAdminChanged(uint64,uint64)": TypedContractEvent<
      RoleAdminChangedEvent.InputTuple,
      RoleAdminChangedEvent.OutputTuple,
      RoleAdminChangedEvent.OutputObject
    >;
    RoleAdminChanged: TypedContractEvent<
      RoleAdminChangedEvent.InputTuple,
      RoleAdminChangedEvent.OutputTuple,
      RoleAdminChangedEvent.OutputObject
    >;

    "RoleGrantDelayChanged(uint64,uint32,uint48)": TypedContractEvent<
      RoleGrantDelayChangedEvent.InputTuple,
      RoleGrantDelayChangedEvent.OutputTuple,
      RoleGrantDelayChangedEvent.OutputObject
    >;
    RoleGrantDelayChanged: TypedContractEvent<
      RoleGrantDelayChangedEvent.InputTuple,
      RoleGrantDelayChangedEvent.OutputTuple,
      RoleGrantDelayChangedEvent.OutputObject
    >;

    "RoleGranted(uint64,address,uint32,uint48,bool)": TypedContractEvent<
      RoleGrantedEvent.InputTuple,
      RoleGrantedEvent.OutputTuple,
      RoleGrantedEvent.OutputObject
    >;
    RoleGranted: TypedContractEvent<
      RoleGrantedEvent.InputTuple,
      RoleGrantedEvent.OutputTuple,
      RoleGrantedEvent.OutputObject
    >;

    "RoleGuardianChanged(uint64,uint64)": TypedContractEvent<
      RoleGuardianChangedEvent.InputTuple,
      RoleGuardianChangedEvent.OutputTuple,
      RoleGuardianChangedEvent.OutputObject
    >;
    RoleGuardianChanged: TypedContractEvent<
      RoleGuardianChangedEvent.InputTuple,
      RoleGuardianChangedEvent.OutputTuple,
      RoleGuardianChangedEvent.OutputObject
    >;

    "RoleLabel(uint64,string)": TypedContractEvent<
      RoleLabelEvent.InputTuple,
      RoleLabelEvent.OutputTuple,
      RoleLabelEvent.OutputObject
    >;
    RoleLabel: TypedContractEvent<
      RoleLabelEvent.InputTuple,
      RoleLabelEvent.OutputTuple,
      RoleLabelEvent.OutputObject
    >;

    "RoleRevoked(uint64,address)": TypedContractEvent<
      RoleRevokedEvent.InputTuple,
      RoleRevokedEvent.OutputTuple,
      RoleRevokedEvent.OutputObject
    >;
    RoleRevoked: TypedContractEvent<
      RoleRevokedEvent.InputTuple,
      RoleRevokedEvent.OutputTuple,
      RoleRevokedEvent.OutputObject
    >;

    "TargetAdminDelayUpdated(address,uint32,uint48)": TypedContractEvent<
      TargetAdminDelayUpdatedEvent.InputTuple,
      TargetAdminDelayUpdatedEvent.OutputTuple,
      TargetAdminDelayUpdatedEvent.OutputObject
    >;
    TargetAdminDelayUpdated: TypedContractEvent<
      TargetAdminDelayUpdatedEvent.InputTuple,
      TargetAdminDelayUpdatedEvent.OutputTuple,
      TargetAdminDelayUpdatedEvent.OutputObject
    >;

    "TargetClosed(address,bool)": TypedContractEvent<
      TargetClosedEvent.InputTuple,
      TargetClosedEvent.OutputTuple,
      TargetClosedEvent.OutputObject
    >;
    TargetClosed: TypedContractEvent<
      TargetClosedEvent.InputTuple,
      TargetClosedEvent.OutputTuple,
      TargetClosedEvent.OutputObject
    >;

    "TargetFunctionRoleUpdated(address,bytes4,uint64)": TypedContractEvent<
      TargetFunctionRoleUpdatedEvent.InputTuple,
      TargetFunctionRoleUpdatedEvent.OutputTuple,
      TargetFunctionRoleUpdatedEvent.OutputObject
    >;
    TargetFunctionRoleUpdated: TypedContractEvent<
      TargetFunctionRoleUpdatedEvent.InputTuple,
      TargetFunctionRoleUpdatedEvent.OutputTuple,
      TargetFunctionRoleUpdatedEvent.OutputObject
    >;
  };
}
