// @openzeppelin/contracts/build/contracts/IERC777.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import type {
  BaseContract,
  BigNumberish,
  BytesLike,
  FunctionFragment,
  Result,
  Interface,
  EventFragment,
  AddressLike,
  ContractRunner,
  ContractMethod,
  Listener,
} from "ethers";
import type {
  TypedContractEvent,
  TypedDeferredTopicFilter,
  TypedEventLog,
  TypedLogDescription,
  TypedListener,
  TypedContractMethod,
} from "../../../../../common";

export interface IERC777Interface extends Interface {
  getFunction(
    nameOrSignature:
      | "authorizeOperator"
      | "authorizeOperator(address)"
      | "balanceOf"
      | "balanceOf(address)"
      | "burn"
      | "burn(uint256,bytes)"
      | "defaultOperators"
      | "defaultOperators()"
      | "granularity"
      | "granularity()"
      | "isOperatorFor"
      | "isOperatorFor(address,address)"
      | "name"
      | "name()"
      | "operatorBurn"
      | "operatorBurn(address,uint256,bytes,bytes)"
      | "operatorSend"
      | "operatorSend(address,address,uint256,bytes,bytes)"
      | "revokeOperator"
      | "revokeOperator(address)"
      | "send"
      | "send(address,uint256,bytes)"
      | "symbol"
      | "symbol()"
      | "totalSupply"
      | "totalSupply()"
  ): FunctionFragment;

  getEvent(
    nameOrSignatureOrTopic:
      | "AuthorizedOperator"
      | "AuthorizedOperator(address,address)"
      | "Burned"
      | "Burned(address,address,uint256,bytes,bytes)"
      | "Minted"
      | "Minted(address,address,uint256,bytes,bytes)"
      | "RevokedOperator"
      | "RevokedOperator(address,address)"
      | "Sent"
      | "Sent(address,address,address,uint256,bytes,bytes)"
  ): EventFragment;

  encodeFunctionData(
    functionFragment: "authorizeOperator",
    values: [AddressLike]
  ): string;
  encodeFunctionData(
    functionFragment: "authorizeOperator(address)",
    values: [AddressLike]
  ): string;
  encodeFunctionData(
    functionFragment: "balanceOf",
    values: [AddressLike]
  ): string;
  encodeFunctionData(
    functionFragment: "balanceOf(address)",
    values: [AddressLike]
  ): string;
  encodeFunctionData(
    functionFragment: "burn",
    values: [BigNumberish, BytesLike]
  ): string;
  encodeFunctionData(
    functionFragment: "burn(uint256,bytes)",
    values: [BigNumberish, BytesLike]
  ): string;
  encodeFunctionData(
    functionFragment: "defaultOperators",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "defaultOperators()",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "granularity",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "granularity()",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "isOperatorFor",
    values: [AddressLike, AddressLike]
  ): string;
  encodeFunctionData(
    functionFragment: "isOperatorFor(address,address)",
    values: [AddressLike, AddressLike]
  ): string;
  encodeFunctionData(functionFragment: "name", values?: undefined): string;
  encodeFunctionData(functionFragment: "name()", values?: undefined): string;
  encodeFunctionData(
    functionFragment: "operatorBurn",
    values: [AddressLike, BigNumberish, BytesLike, BytesLike]
  ): string;
  encodeFunctionData(
    functionFragment: "operatorBurn(address,uint256,bytes,bytes)",
    values: [AddressLike, BigNumberish, BytesLike, BytesLike]
  ): string;
  encodeFunctionData(
    functionFragment: "operatorSend",
    values: [AddressLike, AddressLike, BigNumberish, BytesLike, BytesLike]
  ): string;
  encodeFunctionData(
    functionFragment: "operatorSend(address,address,uint256,bytes,bytes)",
    values: [AddressLike, AddressLike, BigNumberish, BytesLike, BytesLike]
  ): string;
  encodeFunctionData(
    functionFragment: "revokeOperator",
    values: [AddressLike]
  ): string;
  encodeFunctionData(
    functionFragment: "revokeOperator(address)",
    values: [AddressLike]
  ): string;
  encodeFunctionData(
    functionFragment: "send",
    values: [AddressLike, BigNumberish, BytesLike]
  ): string;
  encodeFunctionData(
    functionFragment: "send(address,uint256,bytes)",
    values: [AddressLike, BigNumberish, BytesLike]
  ): string;
  encodeFunctionData(functionFragment: "symbol", values?: undefined): string;
  encodeFunctionData(functionFragment: "symbol()", values?: undefined): string;
  encodeFunctionData(
    functionFragment: "totalSupply",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "totalSupply()",
    values?: undefined
  ): string;

  decodeFunctionResult(
    functionFragment: "authorizeOperator",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "authorizeOperator(address)",
    data: BytesLike
  ): Result;
  decodeFunctionResult(functionFragment: "balanceOf", data: BytesLike): Result;
  decodeFunctionResult(
    functionFragment: "balanceOf(address)",
    data: BytesLike
  ): Result;
  decodeFunctionResult(functionFragment: "burn", data: BytesLike): Result;
  decodeFunctionResult(
    functionFragment: "burn(uint256,bytes)",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "defaultOperators",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "defaultOperators()",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "granularity",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "granularity()",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "isOperatorFor",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "isOperatorFor(address,address)",
    data: BytesLike
  ): Result;
  decodeFunctionResult(functionFragment: "name", data: BytesLike): Result;
  decodeFunctionResult(functionFragment: "name()", data: BytesLike): Result;
  decodeFunctionResult(
    functionFragment: "operatorBurn",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "operatorBurn(address,uint256,bytes,bytes)",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "operatorSend",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "operatorSend(address,address,uint256,bytes,bytes)",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "revokeOperator",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "revokeOperator(address)",
    data: BytesLike
  ): Result;
  decodeFunctionResult(functionFragment: "send", data: BytesLike): Result;
  decodeFunctionResult(
    functionFragment: "send(address,uint256,bytes)",
    data: BytesLike
  ): Result;
  decodeFunctionResult(functionFragment: "symbol", data: BytesLike): Result;
  decodeFunctionResult(functionFragment: "symbol()", data: BytesLike): Result;
  decodeFunctionResult(
    functionFragment: "totalSupply",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "totalSupply()",
    data: BytesLike
  ): Result;
}

export namespace AuthorizedOperatorEvent {
  export type InputTuple = [operator: AddressLike, tokenHolder: AddressLike];
  export type OutputTuple = [operator: string, tokenHolder: string];
  export interface OutputObject {
    operator: string;
    tokenHolder: string;
  }
  export type Event = TypedContractEvent<InputTuple, OutputTuple, OutputObject>;
  export type Filter = TypedDeferredTopicFilter<Event>;
  export type Log = TypedEventLog<Event>;
  export type LogDescription = TypedLogDescription<Event>;
}

export namespace BurnedEvent {
  export type InputTuple = [
    operator: AddressLike,
    from: AddressLike,
    amount: BigNumberish,
    data: BytesLike,
    operatorData: BytesLike
  ];
  export type OutputTuple = [
    operator: string,
    from: string,
    amount: bigint,
    data: string,
    operatorData: string
  ];
  export interface OutputObject {
    operator: string;
    from: string;
    amount: bigint;
    data: string;
    operatorData: string;
  }
  export type Event = TypedContractEvent<InputTuple, OutputTuple, OutputObject>;
  export type Filter = TypedDeferredTopicFilter<Event>;
  export type Log = TypedEventLog<Event>;
  export type LogDescription = TypedLogDescription<Event>;
}

export namespace MintedEvent {
  export type InputTuple = [
    operator: AddressLike,
    to: AddressLike,
    amount: BigNumberish,
    data: BytesLike,
    operatorData: BytesLike
  ];
  export type OutputTuple = [
    operator: string,
    to: string,
    amount: bigint,
    data: string,
    operatorData: string
  ];
  export interface OutputObject {
    operator: string;
    to: string;
    amount: bigint;
    data: string;
    operatorData: string;
  }
  export type Event = TypedContractEvent<InputTuple, OutputTuple, OutputObject>;
  export type Filter = TypedDeferredTopicFilter<Event>;
  export type Log = TypedEventLog<Event>;
  export type LogDescription = TypedLogDescription<Event>;
}

export namespace RevokedOperatorEvent {
  export type InputTuple = [operator: AddressLike, tokenHolder: AddressLike];
  export type OutputTuple = [operator: string, tokenHolder: string];
  export interface OutputObject {
    operator: string;
    tokenHolder: string;
  }
  export type Event = TypedContractEvent<InputTuple, OutputTuple, OutputObject>;
  export type Filter = TypedDeferredTopicFilter<Event>;
  export type Log = TypedEventLog<Event>;
  export type LogDescription = TypedLogDescription<Event>;
}

export namespace SentEvent {
  export type InputTuple = [
    operator: AddressLike,
    from: AddressLike,
    to: AddressLike,
    amount: BigNumberish,
    data: BytesLike,
    operatorData: BytesLike
  ];
  export type OutputTuple = [
    operator: string,
    from: string,
    to: string,
    amount: bigint,
    data: string,
    operatorData: string
  ];
  export interface OutputObject {
    operator: string;
    from: string;
    to: string;
    amount: bigint;
    data: string;
    operatorData: string;
  }
  export type Event = TypedContractEvent<InputTuple, OutputTuple, OutputObject>;
  export type Filter = TypedDeferredTopicFilter<Event>;
  export type Log = TypedEventLog<Event>;
  export type LogDescription = TypedLogDescription<Event>;
}

export interface IERC777 extends BaseContract {
  connect(runner?: ContractRunner | null): IERC777;
  waitForDeployment(): Promise<this>;

  interface: IERC777Interface;

  queryFilter<TCEvent extends TypedContractEvent>(
    event: TCEvent,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TypedEventLog<TCEvent>>>;
  queryFilter<TCEvent extends TypedContractEvent>(
    filter: TypedDeferredTopicFilter<TCEvent>,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TypedEventLog<TCEvent>>>;

  on<TCEvent extends TypedContractEvent>(
    event: TCEvent,
    listener: TypedListener<TCEvent>
  ): Promise<this>;
  on<TCEvent extends TypedContractEvent>(
    filter: TypedDeferredTopicFilter<TCEvent>,
    listener: TypedListener<TCEvent>
  ): Promise<this>;

  once<TCEvent extends TypedContractEvent>(
    event: TCEvent,
    listener: TypedListener<TCEvent>
  ): Promise<this>;
  once<TCEvent extends TypedContractEvent>(
    filter: TypedDeferredTopicFilter<TCEvent>,
    listener: TypedListener<TCEvent>
  ): Promise<this>;

  listeners<TCEvent extends TypedContractEvent>(
    event: TCEvent
  ): Promise<Array<TypedListener<TCEvent>>>;
  listeners(eventName?: string): Promise<Array<Listener>>;
  removeAllListeners<TCEvent extends TypedContractEvent>(
    event?: TCEvent
  ): Promise<this>;

  authorizeOperator: TypedContractMethod<
    [operator: AddressLike],
    [void],
    "nonpayable"
  >;

  "authorizeOperator(address)": TypedContractMethod<
    [operator: AddressLike],
    [void],
    "nonpayable"
  >;

  balanceOf: TypedContractMethod<[owner: AddressLike], [bigint], "view">;

  "balanceOf(address)": TypedContractMethod<
    [owner: AddressLike],
    [bigint],
    "view"
  >;

  burn: TypedContractMethod<
    [amount: BigNumberish, data: BytesLike],
    [void],
    "nonpayable"
  >;

  "burn(uint256,bytes)": TypedContractMethod<
    [amount: BigNumberish, data: BytesLike],
    [void],
    "nonpayable"
  >;

  defaultOperators: TypedContractMethod<[], [string[]], "view">;

  "defaultOperators()": TypedContractMethod<[], [string[]], "view">;

  granularity: TypedContractMethod<[], [bigint], "view">;

  "granularity()": TypedContractMethod<[], [bigint], "view">;

  isOperatorFor: TypedContractMethod<
    [operator: AddressLike, tokenHolder: AddressLike],
    [boolean],
    "view"
  >;

  "isOperatorFor(address,address)": TypedContractMethod<
    [operator: AddressLike, tokenHolder: AddressLike],
    [boolean],
    "view"
  >;

  name: TypedContractMethod<[], [string], "view">;

  "name()": TypedContractMethod<[], [string], "view">;

  operatorBurn: TypedContractMethod<
    [
      account: AddressLike,
      amount: BigNumberish,
      data: BytesLike,
      operatorData: BytesLike
    ],
    [void],
    "nonpayable"
  >;

  "operatorBurn(address,uint256,bytes,bytes)": TypedContractMethod<
    [
      account: AddressLike,
      amount: BigNumberish,
      data: BytesLike,
      operatorData: BytesLike
    ],
    [void],
    "nonpayable"
  >;

  operatorSend: TypedContractMethod<
    [
      sender: AddressLike,
      recipient: AddressLike,
      amount: BigNumberish,
      data: BytesLike,
      operatorData: BytesLike
    ],
    [void],
    "nonpayable"
  >;

  "operatorSend(address,address,uint256,bytes,bytes)": TypedContractMethod<
    [
      sender: AddressLike,
      recipient: AddressLike,
      amount: BigNumberish,
      data: BytesLike,
      operatorData: BytesLike
    ],
    [void],
    "nonpayable"
  >;

  revokeOperator: TypedContractMethod<
    [operator: AddressLike],
    [void],
    "nonpayable"
  >;

  "revokeOperator(address)": TypedContractMethod<
    [operator: AddressLike],
    [void],
    "nonpayable"
  >;

  send: TypedContractMethod<
    [recipient: AddressLike, amount: BigNumberish, data: BytesLike],
    [void],
    "nonpayable"
  >;

  "send(address,uint256,bytes)": TypedContractMethod<
    [recipient: AddressLike, amount: BigNumberish, data: BytesLike],
    [void],
    "nonpayable"
  >;

  symbol: TypedContractMethod<[], [string], "view">;

  "symbol()": TypedContractMethod<[], [string], "view">;

  totalSupply: TypedContractMethod<[], [bigint], "view">;

  "totalSupply()": TypedContractMethod<[], [bigint], "view">;

  getFunction<T extends ContractMethod = ContractMethod>(
    key: string | FunctionFragment
  ): T;

  getFunction(
    nameOrSignature: "authorizeOperator"
  ): TypedContractMethod<[operator: AddressLike], [void], "nonpayable">;
  getFunction(
    nameOrSignature: "authorizeOperator(address)"
  ): TypedContractMethod<[operator: AddressLike], [void], "nonpayable">;
  getFunction(
    nameOrSignature: "balanceOf"
  ): TypedContractMethod<[owner: AddressLike], [bigint], "view">;
  getFunction(
    nameOrSignature: "balanceOf(address)"
  ): TypedContractMethod<[owner: AddressLike], [bigint], "view">;
  getFunction(
    nameOrSignature: "burn"
  ): TypedContractMethod<
    [amount: BigNumberish, data: BytesLike],
    [void],
    "nonpayable"
  >;
  getFunction(
    nameOrSignature: "burn(uint256,bytes)"
  ): TypedContractMethod<
    [amount: BigNumberish, data: BytesLike],
    [void],
    "nonpayable"
  >;
  getFunction(
    nameOrSignature: "defaultOperators"
  ): TypedContractMethod<[], [string[]], "view">;
  getFunction(
    nameOrSignature: "defaultOperators()"
  ): TypedContractMethod<[], [string[]], "view">;
  getFunction(
    nameOrSignature: "granularity"
  ): TypedContractMethod<[], [bigint], "view">;
  getFunction(
    nameOrSignature: "granularity()"
  ): TypedContractMethod<[], [bigint], "view">;
  getFunction(
    nameOrSignature: "isOperatorFor"
  ): TypedContractMethod<
    [operator: AddressLike, tokenHolder: AddressLike],
    [boolean],
    "view"
  >;
  getFunction(
    nameOrSignature: "isOperatorFor(address,address)"
  ): TypedContractMethod<
    [operator: AddressLike, tokenHolder: AddressLike],
    [boolean],
    "view"
  >;
  getFunction(
    nameOrSignature: "name"
  ): TypedContractMethod<[], [string], "view">;
  getFunction(
    nameOrSignature: "name()"
  ): TypedContractMethod<[], [string], "view">;
  getFunction(
    nameOrSignature: "operatorBurn"
  ): TypedContractMethod<
    [
      account: AddressLike,
      amount: BigNumberish,
      data: BytesLike,
      operatorData: BytesLike
    ],
    [void],
    "nonpayable"
  >;
  getFunction(
    nameOrSignature: "operatorBurn(address,uint256,bytes,bytes)"
  ): TypedContractMethod<
    [
      account: AddressLike,
      amount: BigNumberish,
      data: BytesLike,
      operatorData: BytesLike
    ],
    [void],
    "nonpayable"
  >;
  getFunction(
    nameOrSignature: "operatorSend"
  ): TypedContractMethod<
    [
      sender: AddressLike,
      recipient: AddressLike,
      amount: BigNumberish,
      data: BytesLike,
      operatorData: BytesLike
    ],
    [void],
    "nonpayable"
  >;
  getFunction(
    nameOrSignature: "operatorSend(address,address,uint256,bytes,bytes)"
  ): TypedContractMethod<
    [
      sender: AddressLike,
      recipient: AddressLike,
      amount: BigNumberish,
      data: BytesLike,
      operatorData: BytesLike
    ],
    [void],
    "nonpayable"
  >;
  getFunction(
    nameOrSignature: "revokeOperator"
  ): TypedContractMethod<[operator: AddressLike], [void], "nonpayable">;
  getFunction(
    nameOrSignature: "revokeOperator(address)"
  ): TypedContractMethod<[operator: AddressLike], [void], "nonpayable">;
  getFunction(
    nameOrSignature: "send"
  ): TypedContractMethod<
    [recipient: AddressLike, amount: BigNumberish, data: BytesLike],
    [void],
    "nonpayable"
  >;
  getFunction(
    nameOrSignature: "send(address,uint256,bytes)"
  ): TypedContractMethod<
    [recipient: AddressLike, amount: BigNumberish, data: BytesLike],
    [void],
    "nonpayable"
  >;
  getFunction(
    nameOrSignature: "symbol"
  ): TypedContractMethod<[], [string], "view">;
  getFunction(
    nameOrSignature: "symbol()"
  ): TypedContractMethod<[], [string], "view">;
  getFunction(
    nameOrSignature: "totalSupply"
  ): TypedContractMethod<[], [bigint], "view">;
  getFunction(
    nameOrSignature: "totalSupply()"
  ): TypedContractMethod<[], [bigint], "view">;

  getEvent(
    key: "AuthorizedOperator"
  ): TypedContractEvent<
    AuthorizedOperatorEvent.InputTuple,
    AuthorizedOperatorEvent.OutputTuple,
    AuthorizedOperatorEvent.OutputObject
  >;
  getEvent(
    key: "AuthorizedOperator(address,address)"
  ): TypedContractEvent<
    AuthorizedOperator_address_address_Event.InputTuple,
    AuthorizedOperator_address_address_Event.OutputTuple,
    AuthorizedOperator_address_address_Event.OutputObject
  >;
  getEvent(
    key: "Burned"
  ): TypedContractEvent<
    BurnedEvent.InputTuple,
    BurnedEvent.OutputTuple,
    BurnedEvent.OutputObject
  >;
  getEvent(
    key: "Burned(address,address,uint256,bytes,bytes)"
  ): TypedContractEvent<
    Burned_address_address_uint256_bytes_bytes_Event.InputTuple,
    Burned_address_address_uint256_bytes_bytes_Event.OutputTuple,
    Burned_address_address_uint256_bytes_bytes_Event.OutputObject
  >;
  getEvent(
    key: "Minted"
  ): TypedContractEvent<
    MintedEvent.InputTuple,
    MintedEvent.OutputTuple,
    MintedEvent.OutputObject
  >;
  getEvent(
    key: "Minted(address,address,uint256,bytes,bytes)"
  ): TypedContractEvent<
    Minted_address_address_uint256_bytes_bytes_Event.InputTuple,
    Minted_address_address_uint256_bytes_bytes_Event.OutputTuple,
    Minted_address_address_uint256_bytes_bytes_Event.OutputObject
  >;
  getEvent(
    key: "RevokedOperator"
  ): TypedContractEvent<
    RevokedOperatorEvent.InputTuple,
    RevokedOperatorEvent.OutputTuple,
    RevokedOperatorEvent.OutputObject
  >;
  getEvent(
    key: "RevokedOperator(address,address)"
  ): TypedContractEvent<
    RevokedOperator_address_address_Event.InputTuple,
    RevokedOperator_address_address_Event.OutputTuple,
    RevokedOperator_address_address_Event.OutputObject
  >;
  getEvent(
    key: "Sent"
  ): TypedContractEvent<
    SentEvent.InputTuple,
    SentEvent.OutputTuple,
    SentEvent.OutputObject
  >;
  getEvent(
    key: "Sent(address,address,address,uint256,bytes,bytes)"
  ): TypedContractEvent<
    Sent_address_address_address_uint256_bytes_bytes_Event.InputTuple,
    Sent_address_address_address_uint256_bytes_bytes_Event.OutputTuple,
    Sent_address_address_address_uint256_bytes_bytes_Event.OutputObject
  >;

  filters: {
    "AuthorizedOperator(address,address)": TypedContractEvent<
      AuthorizedOperatorEvent.InputTuple,
      AuthorizedOperatorEvent.OutputTuple,
      AuthorizedOperatorEvent.OutputObject
    >;
    AuthorizedOperator: TypedContractEvent<
      AuthorizedOperatorEvent.InputTuple,
      AuthorizedOperatorEvent.OutputTuple,
      AuthorizedOperatorEvent.OutputObject
    >;

    "Burned(address,address,uint256,bytes,bytes)": TypedContractEvent<
      BurnedEvent.InputTuple,
      BurnedEvent.OutputTuple,
      BurnedEvent.OutputObject
    >;
    Burned: TypedContractEvent<
      BurnedEvent.InputTuple,
      BurnedEvent.OutputTuple,
      BurnedEvent.OutputObject
    >;

    "Minted(address,address,uint256,bytes,bytes)": TypedContractEvent<
      MintedEvent.InputTuple,
      MintedEvent.OutputTuple,
      MintedEvent.OutputObject
    >;
    Minted: TypedContractEvent<
      MintedEvent.InputTuple,
      MintedEvent.OutputTuple,
      MintedEvent.OutputObject
    >;

    "RevokedOperator(address,address)": TypedContractEvent<
      RevokedOperatorEvent.InputTuple,
      RevokedOperatorEvent.OutputTuple,
      RevokedOperatorEvent.OutputObject
    >;
    RevokedOperator: TypedContractEvent<
      RevokedOperatorEvent.InputTuple,
      RevokedOperatorEvent.OutputTuple,
      RevokedOperatorEvent.OutputObject
    >;

    "Sent(address,address,address,uint256,bytes,bytes)": TypedContractEvent<
      SentEvent.InputTuple,
      SentEvent.OutputTuple,
      SentEvent.OutputObject
    >;
    Sent: TypedContractEvent<
      SentEvent.InputTuple,
      SentEvent.OutputTuple,
      SentEvent.OutputObject
    >;
  };
}
