// @openzeppelin/contracts/build/contracts/IERC1363Receiver.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import type {
  BaseContract,
  BigNumberish,
  BytesLike,
  FunctionFragment,
  Result,
  Interface,
  AddressLike,
  ContractRunner,
  ContractMethod,
  Listener,
} from "ethers";
import type {
  TypedContractEvent,
  TypedDeferredTopicFilter,
  TypedEventLog,
  TypedListener,
  TypedContractMethod,
} from "../../../../../common";

export interface IERC1363ReceiverInterface extends Interface {
  getFunction(
    nameOrSignature:
      | "onTransferReceived"
      | "onTransferReceived(address,address,uint256,bytes)"
  ): FunctionFragment;

  encodeFunctionData(
    functionFragment: "onTransferReceived",
    values: [AddressLike, AddressLike, BigNumberish, BytesLike]
  ): string;
  encodeFunctionData(
    functionFragment: "onTransferReceived(address,address,uint256,bytes)",
    values: [AddressLike, AddressLike, BigNumberish, BytesLike]
  ): string;

  decodeFunctionResult(
    functionFragment: "onTransferReceived",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "onTransferReceived(address,address,uint256,bytes)",
    data: BytesLike
  ): Result;
}

export interface IERC1363Receiver extends BaseContract {
  connect(runner?: ContractRunner | null): IERC1363Receiver;
  waitForDeployment(): Promise<this>;

  interface: IERC1363ReceiverInterface;

  queryFilter<TCEvent extends TypedContractEvent>(
    event: TCEvent,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TypedEventLog<TCEvent>>>;
  queryFilter<TCEvent extends TypedContractEvent>(
    filter: TypedDeferredTopicFilter<TCEvent>,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TypedEventLog<TCEvent>>>;

  on<TCEvent extends TypedContractEvent>(
    event: TCEvent,
    listener: TypedListener<TCEvent>
  ): Promise<this>;
  on<TCEvent extends TypedContractEvent>(
    filter: TypedDeferredTopicFilter<TCEvent>,
    listener: TypedListener<TCEvent>
  ): Promise<this>;

  once<TCEvent extends TypedContractEvent>(
    event: TCEvent,
    listener: TypedListener<TCEvent>
  ): Promise<this>;
  once<TCEvent extends TypedContractEvent>(
    filter: TypedDeferredTopicFilter<TCEvent>,
    listener: TypedListener<TCEvent>
  ): Promise<this>;

  listeners<TCEvent extends TypedContractEvent>(
    event: TCEvent
  ): Promise<Array<TypedListener<TCEvent>>>;
  listeners(eventName?: string): Promise<Array<Listener>>;
  removeAllListeners<TCEvent extends TypedContractEvent>(
    event?: TCEvent
  ): Promise<this>;

  onTransferReceived: TypedContractMethod<
    [
      operator: AddressLike,
      from: AddressLike,
      amount: BigNumberish,
      data: BytesLike
    ],
    [string],
    "nonpayable"
  >;

  "onTransferReceived(address,address,uint256,bytes)": TypedContractMethod<
    [
      operator: AddressLike,
      from: AddressLike,
      amount: BigNumberish,
      data: BytesLike
    ],
    [string],
    "nonpayable"
  >;

  getFunction<T extends ContractMethod = ContractMethod>(
    key: string | FunctionFragment
  ): T;

  getFunction(
    nameOrSignature: "onTransferReceived"
  ): TypedContractMethod<
    [
      operator: AddressLike,
      from: AddressLike,
      amount: BigNumberish,
      data: BytesLike
    ],
    [string],
    "nonpayable"
  >;
  getFunction(
    nameOrSignature: "onTransferReceived(address,address,uint256,bytes)"
  ): TypedContractMethod<
    [
      operator: AddressLike,
      from: AddressLike,
      amount: BigNumberish,
      data: BytesLike
    ],
    [string],
    "nonpayable"
  >;

  filters: {};
}
